xgcm/grid.py

```diff
@@ -1162,28 +1211,79 @@ def integrate(self, da, axis, **kwargs):
         da_i : xarray.DataArray
             The integrated data
         """
-        check_axes = [ax for ax in axis if ax not in self.axes]
-        check_axes_str = ",".join(check_axes)
-        if any(check_axes):
-            raise ValueError("Axis %s not found in grid object" % check_axes_str)
 
-        dim = []
+        weight = self.get_metric(da, axis)
+        weighted = da * weight
+        # TODO: We should integrate xarray.weighted once available.
+
+        # get dimension(s) corresponding
+        # to `da` and `axis` input
+        dim = self._get_dims_from_axis(da, axis)
+
+        return weighted.sum(dim)
+
+    @docstrings.dedent
+    # !!!! there is a problem here...I cant apply cumsum over multiple axes...
+    def cumint(self, da, axis, **kwargs):
+        """
+        Perform cumulative integral along specified axis or axes,
+        accounting for grid metrics. (e.g. cell length, area, volume)
+
+        Parameters
+        ----------
+        axis : str, list of str
+            Name of the axis on which to act
+        %(neighbor_binary_func.parameters.no_f)s
+
+        Returns
+        -------
+        da_i : xarray.DataArray
+            The cumulatively integrated data
+        """
+
+        weight = self.get_metric(da, axis)
+        weighted = da * weight
+        # TODO: We should integrate xarray.weighted once available.
+
+        # This is a workaround, that should be eliminated
+        # once cumsum can axcept multiple axes
+
+        if isinstance(axis, str):
+            axis = [axis]
+
+        out = weighted
         for ax in axis:
-            all_dim = self.axes[ax].coords.values()
-            matching_dim = [di for di in all_dim if di in da.dims]
-            if len(matching_dim) == 1:
-                dim.append(matching_dim[0])
-            else:
-                raise ValueError(
-                    "Did not find single matching dimension corresponding to axis %s. Got (%s)"
-                    % (ax, matching_dim)
-                )
+            out = self.cumsum(out, ax, **kwargs)
+
+        return out
+
+    @docstrings.dedent
+    def average(self, da, axis, **kwargs):
+        """
+        Perform weighted average along specified axis or axes,
+        accounting for grid metrics. (e.g. cell length, area, volume)
+
+        Parameters
+        ----------
+        axis : str, list of str
+            Name of the axis on which to act
+        %(neighbor_binary_func.parameters.no_f)s
+
+        Returns
+        -------
+        da_i : xarray.DataArray
+            The averaged data
+        """
 
         weight = self.get_metric(da, axis)
         weighted = da * weight
         # TODO: We should integrate xarray.weighted once available.
 
-        return weighted.sum(dim)
+        # get dimension(s) corresponding
+        # to `da` and `axis` input
+        dim = self._get_dims_from_axis(da, axis)
+        # do we need to pass kwargs?
```

Comment:
What kind of kwargs would be passed here? Since it is a reducing operation, the boundaries should not matter?

Reply:
Good point. Ill add that.